### ISSUE TITLE
Cookies not setting properly supabase ssr

### ISSUE THREAD

**Issue by Sof2222** [supabase/ssr#36]
# Bug report

I have already checked and cant see the same issue.
- [x ] I confirm this is a bug with Supabase, not with my own application. 
- [x ] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

I am using supabase-ssr package to log on.

I thought this was only an issue in dev mode as when I ran build mode on Friday it worked, but perhaps I had not properly deleted the cookie when I was testing so am getting the error again now.

Basically the auth-token cookie is not setting properly. If I log on twice, it sets but the first time i log on only sb-__-auth-token-code-verifier is set.
I am unsure if it is something on my side which is causing the error or if there is something timing out in the setting of the second cookie. My code is below.
Note I am using a otp sent to emails for this.

## To Reproduce

This is to get the code:
```
export async function signuplogin(prevState: any, formData: FormData) {
  console.log(formData);
  const validatedFields = schema.safeParse({
    email: formData.get("email"),
  });
  console.log(validatedFields);

  if (!validatedFields.success) {
    console.log(validatedFields.error.flatten().fieldErrors.email);
    return {
      message: validatedFields.error.flatten().fieldErrors,
    };
  }
  console.log(formData);
  const email = formData.get("email") as string;
  return signInOTP({ email });

}

```

This is the server component for createClient:

```
import { createServerClient, type CookieOptions } from "@supabase/ssr";
import { cookies } from "next/headers";

export const createClient = (cookieStore: ReturnType<typeof cookies>) => {
  return createServerClient<Database>(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value, ...options });
          } catch (error) {
            // The `set` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
        remove(name: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value: "", ...options });
          } catch (error) {
            // The `delete` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
      },
    }
  );
};
```

This is to check the OTP code:

```
export async function precheckOTP(prevState: any, formData: FormData) {
  console.log(formData);
  const validatedFields = schema.safeParse({
    code: formData.get("code"),
  });
  console.log(validatedFields);

  if (!validatedFields.success) {
    console.log(validatedFields.error.flatten().fieldErrors.code);
    return {
      message: validatedFields.error.flatten().fieldErrors,
    };
  }
  console.log(formData);
  const code = formData.get("code") as string;
  const user = formData.get("user") as string;
  return checkOTP({ token: code, user: user });
}


const checkOTP = async ({ token, user }: { token: string; user: string }) => {
  const cookieStore = cookies();
  const supabase = createClient(cookieStore);
  const email = atob(user);

  try {
    const { error } = await supabase.auth.verifyOtp({
      email,
      token,
      type: "email",
      options: {
        redirectTo: "/dashboard",
      },
    });

    if (error) {
      return {
        message: { error: "Something went wrong. Please try again." },
      };
    }
  } catch (error) {
    console.log(error);
    return {
      message: { error: "Something went wrong. Please try again." },
    };
  }

  return redirect("/dashboard");
};
```

I am redirected to the dashboard.

However the cookies are not being set properly. The first time:

sb-___-auth-token-code-verifier is set properly.
The second time I log on sb-___-auth-token is set

(Note this is called when someone is on protected:

```
export async function updateSession(request: NextRequest) {
  let response = NextResponse.next({
    request: {
      headers: request.headers,
    },
  });

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return request.cookies.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          request.cookies.set({
            name,
            value,
            ...options,
          });
          response = NextResponse.next({
            request: {
              headers: request.headers,
            },
          });
          response.cookies.set({
            name,
            value,
            ...options,
          });
        },
        remove(name: string, options: CookieOptions) {
          request.cookies.set({
            name,
            value: "",
            ...options,
          });
          response = NextResponse.next({
            request: {
              headers: request.headers,
            },
          });
          response.cookies.set({
            name,
            value: "",
            ...options,
          });
        },
      },
    }
  );

  await supabase.auth.getUser();

  return { supabase, response };
}

```

I have tried with our without this:   await supabase.auth.getUser();

But then what happens is I get thrown from the route a moment later or if I try to navigate and I am thrown out of the protected route.
I then have to log in again in which case the second cookie is set.

## Expected behavior

That the cookies would all set in the first instance and the user is not required to log on twice for them to set

## System information

- OS: [e.g. macOS,]
- Browser (if applies) [e.g., safari and firefox]
- Version of supabase-js: [e.g. 2.42.3] (ssr 0.0.10)
- Version of Node.js: [e.g. 18.17.0]

nextjs version - 14.1.4

**Comment by j4w8n** [comment]
I'm not familiar with Next, but has anyone tried migrating to the new ssr 0.4.x version and using the setAll and getAll methods? It's supposed to better handle Next quirkiness. New docs are available [here](https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app)

**Comment by IHIutch** [comment]
Since this seems related, I'm also sharing here https://github.com/supabase/supabase-js/issues/1202#issuecomment-2220909186

As you can see, this Next starter works, while the Remix starter does not. 

Logging in seems to work fine, but once the cookies are set/read(?) something breaks and the user can never reach `/protected`. One thing to note however, when I was testing locally, once my cookies were set in my Next app, they were able to be read when I started up my Remix app. 

This seems to track with the initial claim that cookies aren't being set correctly, because they appear to be read fine.

**Comment by Sof2222** [comment]
For anyone having the issues - I think it is a nextjs issue. Updating nextjs solved it and I have not had this issue the past week (have been testing)

**Comment by j4w8n** [comment]
> For anyone having the issues - I think it is a nextjs issue. Updating nextjs solved it and I have not had this issue the past week (have been testing)

To what version?

**Comment by lukahukur** [comment]
> For anyone having the issues - I think it is a nextjs issue. Updating nextjs solved it and I have not had this issue the past week (have been testing)

Ye, for what version?

**Comment by NotAProton** [comment]
Thanks a lot @Sof2222!

For reference I upgraded from `"next": "14.2.4",` to `"next": "14.2.5",` and it fixed the issue

**Comment by eposha** [comment]
Same issue, any solutions? Next.js 14.2.5 doen't  resolve this problem for me

**Comment by cdgn-coding** [comment]
We are experiencing this exact same issue. Next version 14.2.5.

**Comment by eposha** [comment]
In my case, I used `export const dynamic = 'force-dynamic'` in a layout that blocked cookies on the first render.

**Comment by mattrigg9** [comment]
This doesn't directly solve OP's problem, but just throwing this out there in case anyone's banging their head against the table. Make sure you don't have any logout links inadvertently getting prefetched. My supabase auth cookie was getting assigned correctly, but then NextJS would prefetch the logout link on my account page, causing the server to immediately delete the cookie.

**Comment by AndrewLester** [comment]
In the password reset case, with Next.js pages router on 14.2.5 and most recent versions of @supabase/ssr and @supabase/auth-js (at time of this post), I have this workaround:
```ts
// @filename: src/pages/api/auth/confirm.ts
import { type EmailOtpType } from '@supabase/supabase-js';
import type { NextApiRequest, NextApiResponse } from 'next';

import { createClient } from '@/lib/utils/supabase/api';
import { serializeCookieHeader, stringToBase64URL } from '@supabase/ssr';

function stringOrFirstString(item: string | string[] | undefined) {
	return Array.isArray(item) ? item[0] : item;
}

export default async function handler(
	req: NextApiRequest,
	res: NextApiResponse,
) {
	if (req.method !== 'GET') {
		res.status(405).appendHeader('Allow', 'GET').end();
		return;
	}

	const queryParams = req.query;
	const token_hash = stringOrFirstString(queryParams.token_hash);
	const type = stringOrFirstString(queryParams.type);

	let next = '/error';

	if (token_hash && type) {
		const supabase = createClient(req, res);
		const { error, data } = await supabase.auth.verifyOtp({
			type: type as EmailOtpType,
			token_hash,
		});
		if (error) {
			console.error(error);
		} else {
			// Workaround for https://github.com/supabase/ssr/issues/36
			const supabaseURL = new URL(process.env.NEXT_PUBLIC_SUPABASE_URL!);
			const supabaseCookieDomainPrefix = supabaseURL.hostname.split('.')[0];
			const authTokenCookie = {
				name: `sb-${supabaseCookieDomainPrefix}-auth-token`,
				value: `base64-${stringToBase64URL(JSON.stringify(data.session!))}`,
				options: {
					path: '/',
					sameSite: 'lax',
					httpOnly: false,
					maxAge: 31536000000,
				},
			};
			res.setHeader(
				'Set-Cookie',
				[authTokenCookie].map(({ name, value, options }) =>
					serializeCookieHeader(name, value, options),
				),
			);
			next = stringOrFirstString(queryParams.next) || '/';
		}
	}

	res.redirect(next);
}
```

Maybe I've configured something wrong, but this if statement never runs the if branch because setItem is never called with the code-verifier cookie as the key:
https://github.com/supabase/ssr/blob/ae6af1d3f9a405dc8251b595a6b3a3c9d56b9b7f/src/cookies.ts#L358

**Comment by KedalenDev** [comment]
> This doesn't directly solve OP's problem, but just throwing this out there in case anyone's banging their head against the table. Make sure you don't have any logout links inadvertently getting prefetched. My supabase auth cookie was getting assigned correctly, but then NextJS would prefetch the logout link on my account page, causing the server to immediately delete the cookie.

God thank you, I rarely use NEXTJS and I always forget to remove the prefetch behaviour

**Comment by denull0** [comment]
Same problem when I do supabase from a route. Surpsied that this is not revolsed after so many months.

**Comment by benrobertsonio** [comment]
@mattrigg9 you are a hero. thank you!

**Comment by ksloan** [comment]
@mattrigg9 I got caught by the exact same thing, thanks for the tip.

In my case this was only happening in production since ["Prefetching is not enabled in development, only in production."](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#2-prefetching)

**Comment by Aoi-Takahashi** [comment]
Why is everyone using Next.js...? I am having the same problem with Remix　：（

**Comment by foadgr** [comment]
tldr; it may be worth checking for version incompatibilities between the Supabase libraries being used.

i had this issue and resolved by making sure all supabase actions were using @supabase/ssr. It's possible you are using @supabase/auth-helpers-nextjs elsewhere in your app. Improper setting in the cookie likely stems from using different supabase client creation methods in different parts of your app.

**Comment by AdzeB** [comment]
I have the same problem and I am using the latest version of the superbase client and also not using  @supabase/auth-helpers-nextjs... My Quick fix(which still doesn't work for all my users) is to set the cookie manually

**Comment by jbro-io** [comment]
> This doesn't directly solve OP's problem, but just throwing this out there in case anyone's banging their head against the table. Make sure you don't have any logout links inadvertently getting prefetched. My supabase auth cookie was getting assigned correctly, but then NextJS would prefetch the logout link on my account page, causing the server to immediately delete the cookie.

You sir are a godsend, thank you!

**Comment by simon-marcus** [comment]
I'm on Next 14.2.20 (the latest version of Next 14). After mildly excruciating debugging, I realized that upgrading the `@supabase/ssr` package from `0.3.0` to 0.5.0` introduced a variety of auth issues that would remove the cookie from the browser. Upgrading to `@ssr@0.52` appears to have fixed the issue.

**Comment by rishabhsxn** [comment]
> logout

Can someone explain this?
I'm new to NextJs and facing the same issue.
When I call supabaseClient.auth.signup(), the cookies are not being set (only the auth-token-code-verifier cookie is set)
But when I login using email and password, the auth-token cookie is successfully set.

I'm not sure if I'm facing this issue because of the mentioned reason of pre-fetching (I'm facing this issue in local development)

**Comment by rishabhsxn** [comment]
> This doesn't directly solve OP's problem, but just throwing this out there in case anyone's banging their head against the table. Make sure you don't have any logout links inadvertently getting prefetched. My supabase auth cookie was getting assigned correctly, but then NextJS would prefetch the logout link on my account page, causing the server to immediately delete the cookie.

Can someone explain this?
I'm new to NextJs and facing the same issue.
When I call supabaseClient.auth.signup(), the cookies are not being set (only the auth-token-code-verifier cookie is set)
But when I login using email and password, the auth-token cookie is successfully set.

I'm not sure if I'm facing this issue because of the mentioned reason of pre-fetching (I'm facing this issue in local development)

**Comment by rishabhsxn** [comment]
Nevermind.
Found the root cause.
I had `Confirm email` setting on in the Auth Providers -> Email

When this setting is ON, the auth-token cookie won't be created unless the user confirms the email via the Confirmation link

**Comment by PatrickHallek** [comment]
While this might not be the exact issue for the OP, here is a reminder to double-check your host redirections.

I experienced very similar symptoms in my local application, and the root cause turned out to be surprisingly simple. In my callback hook, I was redirecting to `http://localhost:3000` instead of `http://127.0.0.1:3000`.

As a result, the session tokens were being set correctly on `127.0.0.1:3000`, but after the redirect to `localhost:3000`, I had to log in again. This worked because the origin changed to localhost, so a separate session was initiated.

**Comment by handlebauer** [comment]
My solution was similar to @PatrickHallek's

Needed to update the values for `site_url` and `additional_redirect_urls` to use `localhost` in my`config.toml`

**Comment by w7tf** [comment]
Might not be the same issue as OP but in my case it was the streaming API that was causing the issue.

I am using tRPC in my case I had to switch from unstable_httpBatchStreamLink to httpBatchLink

**If you're not using tRPC** check if any streaming API within your framework is causing the issue.

**Comment by JorensM** [comment]
My silly mistake was that I was signing the user in from the server via server actions. I used the Supabase client from the client side and everything started working.

**Comment by gtmetric** [comment]
I encountered a similar issue. In my reset-password flow, I got the correct values in `request.cookies` in my API endpoint, but after redirecting to a page, the cookies were gone. I managed to solve this by setting the `request.cookies` to the Next.js cookie store manually.

**Comment by DanielGiljam** [comment]
I had this issue just now when trying to set up [Password-Based Auth](https://supabase.com/ui/docs/nextjs/password-based-auth) from the (quite recent, I think) Supabase UI Library.

After lots of debugging, I think I figured out the root cause. Not sure if anyone has mentioned it before, didn't bother reading through all the comments and nothing was particularly highlighted as a solution, so I assume it hasn't been mentioned before.

Anyways, what I found is that it came down to how browsers treat the [`Set-Cookie`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Set-Cookie) header in redirect responses.

When I inspected the response from reset password email link (`GET /auth/confirm?token_hash=...`), I found that in the response, the `Set-Cookie` header was set properly. So it looked something like this:

```
GET /auth/confirm?token_hash=...
Host: 127.0.0.1:3000
...

HTTP/1.1 307 Temporary Redirect
Location: http://localhost:3000/auth/update-password
Set-Cookie: sb-127-auth-token=base64-...
...
```

The catch is that because the origin of the original request URL and the redirect response's [`Location`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Location) differ, the browser will ignore the `Set-Cookie` header of the redirect response.

There are cases where browsers would still respect the `Set-Cookie` header in a cross-origin redirect response, e.g. if [`SameSite`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Set-Cookie#samesitesamesite-value) is [`None`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Set-Cookie#none) and [`Secure`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Set-Cookie#secure) is set for the cookie's options. But in this case, the cookie's options (at least in development) were:

```
Path=/; Expires=Thu, 21 May 2026 06:26:18 GMT; Max-Age=34560000; SameSite=lax
```

The solution is to ensure [`SITE_URL`](https://supabase.com/docs/guides/auth/redirect-urls) always is the same origin as the URL that you pass to the `redirectTo` option in [`auth.resetPasswordForEmail`](https://supabase.com/docs/reference/javascript/auth-resetpasswordforemail) (or to the `emailRedirectTo` option in [`auth.signUp`](https://supabase.com/docs/reference/javascript/auth-signup) for that matter).

**Comment by DanielGiljam** [comment]
> While this might not be the exact issue for the OP, here is a reminder to double-check your host redirections.
> 
> I experienced very similar symptoms in my local application, and the root cause turned out to be surprisingly simple. In my callback hook, I was redirecting to `http://localhost:3000` instead of `http://127.0.0.1:3000`.
> 
> As a result, the session tokens were being set correctly on `127.0.0.1:3000`, but after the redirect to `localhost:3000`, I had to log in again.

Okay, this was mentioned before 😅 should read comments more carefully next time. Well, at least for me, I experienced cookies not setting properly when using `@supabase/ssr` and this was the solution!

**Comment by matteopaoli** [comment]
I had a similar issue where the auth cookie was not set, I've realized the problem was i was using two different URLs to connect to my supabase instance between client-side and server-side 

NEXT_PUBLIC_SUPABASE_URL=https://supabase.domain.app // client-side
SERVER_SUPABASE_URL=https://kong:4000


this mismatch caused the cookie to have a different name instead of the one expected

**Comment by mandarini** [comment]
Closing this, since it seems to be fixed. If you find that the issue persists, please tag me and I will check again/reopen.
